src/lua/lua_util.c

```diff
@@ -2470,6 +2489,83 @@ lua_util_is_utf_spoofed (lua_State *L)
 	return nres;
 }
 
+static gint
+lua_util_get_string_stats (lua_State *L)
+{
+	LUA_TRACE_POINT;
+	gsize len_of_string;
+	gint num_of_digits = 0, num_of_letters = 0;
+	const gchar *string_to_check = lua_tolstring (L, 1, &len_of_string);
+
+	if (string_to_check) {
+		while (*string_to_check != '\0') {
+			if (g_ascii_isdigit(*string_to_check)) {
+				num_of_digits++;
+			} else if (g_ascii_isalpha(*string_to_check)) {
+				num_of_letters++;
+			}
+			string_to_check++;
+		}
+	}
+	else {
+		return luaL_error (L, "invalid arguments");
+	}
+
+	lua_createtable(L, 0, 2);
+	lua_pushstring(L, "digits");
+	lua_pushinteger(L, num_of_digits);
+	lua_settable(L, -3);
+	lua_pushstring(L, "letters");
+	lua_pushinteger(L, num_of_letters);
+	lua_settable(L, -3);
+
+	return 1;
+}
+
+
+static gint
+lua_util_is_utf_outside_range(lua_State *L)
+{
+	LUA_TRACE_POINT;
+	gsize len_of_string;
+	gint ret;
+	const gchar *string_to_check = lua_tolstring (L, 1, &len_of_string);
+	guint32 range_start = lua_tointeger (L, 2);
+	guint32 range_end = lua_tointeger (L, 3);
+
+	USpoofChecker *spc_sgl;
+	USet * allowed_chars;
+	UErrorCode uc_err = U_ZERO_ERROR;
+
+	if (string_to_check) {
+		spc_sgl = uspoof_open (&uc_err);
```

Comment:
This can be quite expensive operation. Have you tested performance?

Reply:
I have checked performance and You are right It's 4times slower than similar `util.is_utf_spoofed`. For 1M calls I have 1.3s runtime for `util.is_utf_outside_range` vs 0.3 for `util.is_utf_spoofed`. It's still enough for my purpose.

Problem is that I have no clue how to skip this object initialization on every call and allow customer to use different call parameters at the same time. I think that I cannot use approach from `util.is_utf_spoofed` (static variable). Maybe I should create lua module for this purpose? It will limit the scope of initialization to once per task or maybe You have better idea?

Reply:
Probably, create a static hash table (or `rspamd_lru_hash_t`) that would store some certain amount of entries to speed up generic calls. I suppose that the parameters set would be strictly defined anyway.